### PR TITLE
refactor: don't share `reqwest::HttpClient` across tokio runtimes

### PR DESCRIPTION
**Pull request by dsherret**
This also fixes several issues where we weren't properly creating http clients with the user's settings.

### CODE REVIEW

File: cli/file_fetcher.rs [hunk 1]
```diff
@@ -641,140 +634,17 @@ impl FileFetcher {
   }
 }
 
-#[derive(Debug, Eq, PartialEq)]
-enum FetchOnceResult {
```
Comment (dsherret) [inline cli/file_fetcher.rs:645]: Moved to `HttpClient` so that more of the `reqwest` api usage is buried in there.

File: cli/http_util.rs [hunk 2]
```diff
@@ -407,14 +518,67 @@ pub async fn get_response_body_with_progress(
...
+    let new_url = resolve_url_from_location(request_url, location_string);
+    Ok(new_url)
+  } else {
+    Err(generic_error(format!(
+      "Redirection from '{request_url}' did not provide location header"
+    )))
+  }
+}
+
+#[allow(clippy::print_stdout)]
```
Comment (dsherret) [inline cli/http_util.rs:565]: The below is mostly copy/paste from file_fetcher.rs

File: cli/http_util.rs [hunk 3]
```diff
@@ -407,14 +518,67 @@ pub async fn get_response_body_with_progress(
   Ok(bytes.into())
 }
 
+/// Construct the next uri based on base uri and location header fragment
```
Comment (dsherret) [inline cli/http_util.rs:521]: Copy+pste from file_fetcher.rs

File: cli/tools/installer.rs [hunk 4]
```diff
@@ -133,15 +133,21 @@ fn get_installer_root() -> Result<PathBuf, io::Error> {
...
+pub async fn infer_name_from_url(
+  http_client_provider: &HttpClientProvider,
+  url: &Url,
+) -> Option<String> {
   // If there's an absolute url with no path, eg. https://my-cli.com
   // perform a request, and see if it redirects another file instead.
   let mut url = url.clone();
 
   if url.path() == "/" {
-    let client = HttpClient::new(None, None);
```
Comment (dsherret) [inline cli/tools/installer.rs:142]: Bug. This was not taking into account the user's configuration.

File: cli/tools/registry/provenance.rs [hunk 5]
```diff
@@ -672,9 +682,8 @@ async fn testify(
     },
   };
 
-  let client = Client::new();
```
Comment (dsherret) [inline cli/tools/registry/provenance.rs:675]: Bug.

File: cli/tools/registry/provenance.rs [hunk 6]
```diff
@@ -532,27 +562,6 @@ fn extract_jwt_subject(token: &str) -> Result<String, AnyError> {
...
-async fn gha_request_token(aud: &str) -> Result<String, AnyError> {
-  let Ok(req_url) = env::var("ACTIONS_ID_TOKEN_REQUEST_URL") else {
-    bail!("Not running in GitHub Actions");
-  };
-
-  let Some(token) = gha_oidc_token() else {
-    bail!("No OIDC token available");
-  };
-
-  let client = Client::new();
```
Comment (dsherret) [inline cli/tools/registry/provenance.rs:544]: Bug.

File: cli/clippy.toml
Comment (bartlomieju) [inline cli/clippy.toml:1]: Good thinking

File: cli/http_util.rs [hunk 7]
```diff
@@ -246,77 +231,204 @@ impl HttpClient {
...
-  pub fn from_client(client: reqwest::Client) -> Self {
-    let result = Self {
-      options: Default::default(),
-      root_cert_store_provider: Default::default(),
-      cell: Default::default(),
-    };
-    result.cell.set(client).unwrap();
-    result
+  // todo(dsherret): improve/revamp `HttpClient`'s api in order to
+  // remove this and make `ReqwestClient`
```
Comment (bartlomieju) [inline cli/http_util.rs:239]: Do you mean `ReqwestClient` or `reqwest::Client`? If the latter then I think there's no way around it if we want to disallow sending across threads

**Review by bartlomieju — APPROVED** [review]

**Comment by bartlomieju** [comment]
I'm a bit wary about all these clones, but let's see if we see any regression in the benchmarks.

File: cli/http_util.rs (hunk 7 above)
Comment (dsherret) [reply cli/http_util.rs:239]: Oh, this is an old todo with a previous experimental API I had. I'll remove this.

**Comment by dsherret** [comment]
> I'm a bit wary about all these clones, but let's see if we see any regression in the benchmarks.

Should be fine. It's only cloning Arcs and only when making http requests.

**Pull request merged**